feat(rules): add breaking-change-exclamation-mark

adamchristiansen opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

A breaking change can be denoted by either `!` in the header (e.g., `feat!: add advanced search filters`) or by including `BREAKING CHANGE` in the footer (e.g., `BREAKING CHANGE: old queries no longer work with the new API.`). This pull request adds the `breaking-change-exclamation-mark` rule that requires a `BREAKING CHANGE` footer if `!` appears in the header and a `!` in the header if `BREAKING CHANGE` appears in the footer. The idea is that it is all or nothing, so if one breaking change marker is present the other must be present too.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Implements and closes #4547.

This feature enforces a solution to three problems:

1. Using just `!` might not convey enough information. For example, `feat!: add advanced search filters`. Why is this a breaking change? How do I know what broke unless the footer is present? Using this rule will fail because a footer indicating a breaking change must be added in this case.
2. Using just `BREAKING CHANGE` might cause confusion. For example, if I use something like `git log --online` to quickly look at changes and I see `feat: add advanced search filters`, I have no idea that it is a breaking change without looking at that specific commit in greater detail. This rule will fail because an exclamation mark must be added to the head in this case.
3. This rule enforces the broadest tooling coverage since both means of denoting breaking changes are present in commit messages.

## Usage examples

<!--- Provide examples of intended usage -->

```js
// commitlint.config.js
module.exports = {
  rules: {
    'breaking-change-exclamation-mark': [2, 'always']
  }
};
```

```sh
echo "feat!: add something\n\nBREAKING CHANGE: breaks something" | commitlint # passes
echo "feat: add something" | commitlint # passes
echo "feat!: add something" | commitlint # fails
echo "feat: add something\n\nBREAKING CHANGE: breaks something" | commitlint # fails
```

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->

I added a new test `@commitlint/rules/src/breaking-change-exclamation-mark.test.ts` that follows the structure of existing tests. The commit messages to tests were decided based on equivalence partitions.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. See the README for information on testing. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

JounQin commented on @commitlint/rules/src/breaking-change-exclamation-mark.ts:
```diff
@@ -0,0 +1,35 @@
+		return [true];
+	}
+
+	const hasExclamationMark = !!header && /!:/.test(header);
```
Use `breakingHeaderPattern` accordingly:

https://github.com/conventional-changelog/conventional-changelog/blob/1552de5434163eb63d3bf73d97bcb3bf34de8586/packages/conventional-changelog-conventionalcommits/src/parser.js#L4

JounQin replied:
Personally I think such tiny change is fine to be inclined in one PR.

adamchristiansen replied:
I will make the change.

JounQin commented on @commitlint/rules/src/breaking-change-exclamation-mark.ts:
```diff
@@ -0,0 +1,35 @@
+	}
+
+	const hasExclamationMark = !!header && /!:/.test(header);
+	const hasBreakingChange = !!footer && /BREAKING[ -]CHANGE:/.test(footer);
```
We need make sure the footer line starts with `BREAKING` not only includes?

JounQin commented on @commitlint/rules/src/breaking-change-exclamation-mark.ts:
```diff
@@ -0,0 +1,35 @@
+import message from "@commitlint/message";
+import { SyncRule } from "@commitlint/types";
+
+export const breakingChangeExclamationMark: SyncRule = (
```
The new rule should be added into https://github.com/conventional-changelog/commitlint/blob/0735b631c6899c55a4ce2d94863fad8483c6b619/%40commitlint/types/src/rules.ts#L92

adamchristiansen commented on @commitlint/rules/src/index.ts:
```diff
@@ -36,6 +37,7 @@ import { typeMaxLength } from "./type-max-length.js";
 import { typeMinLength } from "./type-min-length.js";
 
 export default {
+	"breaking-change-exclamation-mark": breakingChangeExclamationMark,
```
Agreed. I will fix the sorting.

adamchristiansen commented on docs/reference/rules.md:
```diff
@@ -1,5 +1,19 @@
 # Rules
 
+## breaking-change-exclamation-mark
```
Agreed again; I will fix this. It has bothered me in the past looking things up on this page as it is about 90% in the right order.

adamchristiansen commented:
I made several pushes because I resolved a conflict with a merge instead of performing a rebase 🤦. I have implemented all suggestions and rebased the branch onto the latest master, so I believe it is ready now.
